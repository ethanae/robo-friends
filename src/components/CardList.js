import React from "react";

//components
import Card from './Card';

const CardList = ({ robots }) => {
    return (
        robots.map(r => {
            return <Card key={r.id} id={r.id} name={r.name} email={r.email} />
        })
    );
}

export default CardList;

