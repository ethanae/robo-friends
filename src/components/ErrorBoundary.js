import React, { Component } from "react";

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hasError: false
    }
  }

  componentDidCatch(error, info) {
    this.setState({ hasError: true });
  }

  render() {
    if (this.state.hasError) {
      console.log('error boundary ***********************88');
      return <h1>Oooops. An error occurred</h1>;
    }
    return this.props.children;
  }
}

export default ErrorBoundary;
