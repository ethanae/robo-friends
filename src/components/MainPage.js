import React, { Component } from "react";

//components
import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Fixed from '../components/Fixed';
import ErrorBoundary from '../components/ErrorBoundary';

class MainPage extends Component {

  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = () => this.props.robots.filter(robot => robot.name.toLowerCase().includes(this.props.searchField.toLowerCase()));

  render() {
    const { onSearchChanged, isPending } = this.props;

    if (isPending)
      return <h1 className='f3 tc ma5'>Loading friends...</h1>;

    return (
      <div className='tc'>
        <Fixed>
          <h1 className='f1'>RoboFriends</h1>
          <SearchBox searchChanged={onSearchChanged} />
        </Fixed>
        <div className='mt7'>
          <ErrorBoundary>
            <CardList robots={this.filterRobots()} />
          </ErrorBoundary>
        </div>
      </div>
    );
  }
}

export default MainPage;