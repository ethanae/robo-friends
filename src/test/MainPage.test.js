import React from 'react';
import MainPage from '../components/MainPage';
import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import ErrorBoundary from '../components/ErrorBoundary';
import { shallow, mount } from 'enzyme';

it('renders MainPage', () => {
  const mockProps = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: '',
    isPending: false
  };
  const wrapper = shallow(<MainPage {...mockProps} />);
  expect(wrapper).toMatchSnapshot();
});

describe('MainPage.filterRobots test', () => {
  it('filters robots - returns array of objects', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [{
        id: 3,
        name: 'JoN',
        email: 'j@j.com'
      }],
      searchField: 'on',
      isPending: false
    };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.instance().filterRobots()).toEqual(mockProps.robots);
  });

  it('filters robots - returns empty array', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [{
        id: 3,
        name: 'JoN',
        email: 'j@j.com'
      }],
      searchField: 'terminator',
      isPending: false
    };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.instance().filterRobots()).toEqual([]);
  });
});

describe('Fetch robots state tests', () => {
  it('displays loading message', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [],
      searchField: '',
      isPending: true
    };
    const wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.find(CardList)).toHaveLength(0);
    expect(wrapper.find(SearchBox)).toHaveLength(0);
    expect(wrapper.find('h1').text()).toEqual('Loading friends...');
    expect(wrapper).toMatchSnapshot();
  });

  it('displays loading message then renders full component', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [],
      searchField: '',
      isPending: true
    };
    const pendingWrapper = shallow(<MainPage {...mockProps} />);
    expect(pendingWrapper.find(CardList)).toHaveLength(0);
    expect(pendingWrapper.find(SearchBox)).toHaveLength(0);
    expect(pendingWrapper.find('h1').text()).toEqual('Loading friends...');

    const resolvedWrapper = shallow(<MainPage {...{ ...mockProps, isPending: false }} />);
    expect(resolvedWrapper.find(CardList)).toHaveLength(1);
    expect(resolvedWrapper.find(SearchBox)).toHaveLength(1);
  });
});

describe('<ErrorBoundary> within <MainPage>', () => {
  it('handles CardList error and invokes ErrorBoundary.componentDidCatch', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [],
      searchField: '',
      isPending: false
    };
    const errorBoundarySpy = jest.spyOn(ErrorBoundary.prototype, 'componentDidCatch');
    const wrapper = mount(<MainPage {...mockProps} />);
    wrapper.find(CardList).simulateError(new Error());
    expect(errorBoundarySpy).toHaveBeenCalled();
  });
});
