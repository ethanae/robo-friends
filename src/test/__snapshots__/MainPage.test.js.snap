// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch robots state tests displays loading message 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainPage
    isPending={true}
    onRequestRobots={[Function]}
    robots={Array []}
    searchField=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Loading friends...",
      "className": "f3 tc ma5",
    },
    "ref": null,
    "rendered": "Loading friends...",
    "type": "h1",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Loading friends...",
        "className": "f3 tc ma5",
      },
      "ref": null,
      "rendered": "Loading friends...",
      "type": "h1",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders MainPage 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainPage
    isPending={false}
    onRequestRobots={[Function]}
    robots={Array []}
    searchField=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Fixed>
          <h1
                    className="f1"
          >
                    RoboFriends
          </h1>
          <SearchBox
                    searchChanged={undefined}
          />
</Fixed>,
        <div
          className="mt7"
>
          <ErrorBoundary>
                    <CardList
                              robots={Array []}
                    />
          </ErrorBoundary>
</div>,
      ],
      "className": "tc",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <h1
              className="f1"
>
              RoboFriends
</h1>,
            <SearchBox
              searchChanged={undefined}
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "RoboFriends",
              "className": "f1",
            },
            "ref": null,
            "rendered": "RoboFriends",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "searchChanged": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ErrorBoundary>
            <CardList
                        robots={Array []}
            />
</ErrorBoundary>,
          "className": "mt7",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <CardList
              robots={Array []}
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "robots": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Fixed>
            <h1
                        className="f1"
            >
                        RoboFriends
            </h1>
            <SearchBox
                        searchChanged={undefined}
            />
</Fixed>,
          <div
            className="mt7"
>
            <ErrorBoundary>
                        <CardList
                                    robots={Array []}
                        />
            </ErrorBoundary>
</div>,
        ],
        "className": "tc",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <h1
                className="f1"
>
                RoboFriends
</h1>,
              <SearchBox
                searchChanged={undefined}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "RoboFriends",
                "className": "f1",
              },
              "ref": null,
              "rendered": "RoboFriends",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "searchChanged": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ErrorBoundary>
              <CardList
                            robots={Array []}
              />
</ErrorBoundary>,
            "className": "mt7",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <CardList
                robots={Array []}
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "robots": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
